### DISCUSSION TITLE
How to make a spline in the C sources

### DISCUSSION THREAD

**Discussion by linusromer** [fontforge/fontforge#4673] [Q&A] [answered]
I am working on an implementation of [Raph Levien's Fitting algorithm](https://raphlinus.github.io/curves/2021/03/11/bezier-fitting.html) for FontForge; see also [typedrawers discussion](https://typedrawers.com/discussion/3975/fitting-cubic-bezier-curves). However, (for testing) I am struggling how to make a spline consisting of two bezier segments am,an,bp,bm and bm,bn,cp,cn.
This is what I have tried, but things like `sab->from->next->to->prev->splines[0].a` seem to point outside the defined quantities:
`BasePoint am = {0,100};
BasePoint an = {27,100};
BasePoint ap = {-10,100};
BasePoint bm = {71,71};
BasePoint bn = {89,53};
BasePoint bp = {53,89};
BasePoint cm = {100,0};
BasePoint cn = {100,-11};
BasePoint cp = {100,28};
SplinePoint a[1] = {{am,an,ap}};
SplinePoint b[1] = {{bm,bn,bp}};
SplinePoint c[1] = {{cm,cn,cp}};
Spline *sab = SplineMake3(a,b);
Spline *sbc = SplineMake3(b,c);`
What I am missing? What is the right way to define such a spline?

**Comment by skef** [comment, marked as answer]
FontForge represents cubic splines in two ways:

1. The on-curve and control points, stored in `SplinePoint` objects (`me` for the on-curve points, `nextcp` and `prevcp` for the control-points.
2. "expanded" x and y cubic polynomial coefficients, with x in splines[0] and y in splines[1]. (See https://fontforge.org/docs/techref/bezier.html ). 

As far as I can tell from your naming convention you're defining your SplinePoints validly. nonextcp and noprevcp should be false, so if you're callocing or memsetting that should be fine.

So is the issue perhaps that you're not expecting the expanded polynomial coefficients?

**Reply by linusromer** [reply]
Well, I actually still do not know where the problem is, but your answer made me consider the first way again and helped me solve my issue. So when running through the points of a spline I am now doing
`for ( frompoint = from, topoint = from->next->to ; ; frompoint = topoint->next->from, topoint = topoint->next->to )`
instead of George's
`for ( np = from->next->to; ; np = np->next->to )`
and it has even a benefit: One does not need as many arrows `->`

**Comment by skef** [comment]
(BTW as long as you're in the sources you might want to glance at #4642 and #4671. Those are upcoming features that might be relevant to your past and/or future FontForge work.)
